### PR TITLE
Upload Updates

### PR DESCRIPTION
**Pull request by JohnDoneth**
## Overview

The original author of the library had some changes that would be nice to have and make using this library easier. I've sifted through the many original changes from the next branch and worked on most of them to completion while omitting some features and adding changes that we require.

## Changes

- Ability to verify uploads are the file type by content instead of trusting the file extension. For example, we check if PDFs are actually PDFs, JPGs, etc.
- Ability to restrict uploads by size / image dimensions / or any other metadata we extract.
- Now saves the original encrypted filename/MD5 checksum/byte size/content-type/metadata in the database.
- Uploads are tracked as a record in a single "blobs" table along with the metadata above so improper deletions of the original record alone will not result in orphaned records in S3
- Now supports arbitrary file transformations/versions with less effort. A version of a blob is just another blob in the blobs table with a foreign key to the original and a "variant" field delineating which version it is.
- Updating of dependencies as well as Erlang/Elixir versions
- Added Credo

### CODE REVIEW

File: lib/upload/blob.ex [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+  end
+
+  defp add_extension_from_mime(changeset) when changeset.valid? do
+    mime = get_field(changeset, :content_type)
+    extension = MIME.extensions(mime) |> List.first()
+
+    if extension do
+      update_change(changeset, :key, fn key ->
+        key <> "." <> extension
+      end)
```
Comment (JohnDoneth) [inline lib/upload/blob.ex:85]: Try uploading image.jpg, see if it creates image.jpg.jpg

File: test/upload/multi_test.exs [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    end
+
+    test "removes variants from the file_store storage" do
+      changeset = change_person(%{avatar: @upload})
+
+      assert {:ok, %{person: person}} = upload_person(changeset)
+
+      {:ok, blob_variant} = Upload.create_variant(person.avatar, "small", &small_transform_avif/3)
+      assert blob_variant.key == "uploads/users/avatars/123/variant/small.avif"
+      assert blob_variant.key in list_uploaded_keys()
```
Comment (JohnDoneth) [inline test/upload/multi_test.exs]: Add 2 variants to test

File: lib/upload.ex [hunk 3]
```diff
@@ -3,207 +3,167 @@ defmodule Upload do
...
-       path: path,
-       filename: filename,
-       status: :pending
-     }}
+  @spec create_variant(Blob.t(), String.t(), any()) :: {:ok, Blob.t()} | {:error, any()}
+  def create_variant(original_blob, variant, transform_fn) when is_function(transform_fn, 3) do
+    repo = Upload.Config.repo()
+    variant = to_string(variant)
+
+    {:ok, multi_result} =
```
Comment (JohnDoneth) [inline lib/upload.ex]: Handle this error correctly.

File: lib/upload.ex [hunk 4]
```diff
@@ -3,207 +3,167 @@ defmodule Upload do
...
+      Ecto.Multi.new()
+      |> Upload.Multi.remove_existing_variant(original_blob, variant)
+      |> Upload.Multi.download_and_insert_variant(original_blob, variant, transform_fn)
+      |> repo.transaction()
+
+    Map.fetch!(multi_result, "download_and_insert_#{variant}")
```
Comment (JohnDoneth) [inline lib/upload.ex]: Maybe use carrot to pattern match

File: lib/upload.ex [hunk 5]
```diff
@@ -3,207 +3,167 @@ defmodule Upload do
...
-    end
+  @spec create_multiple_variants(Blob.t(), [variant_id()], any()) ::
+          {:ok, [Blob.t()]} | {:error, any()}
+  def create_multiple_variants(original_blob, variants, transform_fn)
+      when is_function(transform_fn, 3) do
+    variants = Enum.map(variants, &to_string/1)
+    repo = Upload.Config.repo()
+    multi = Ecto.Multi.new()
+
+    {:ok, multi_result} =
```
Comment (JohnDoneth) [inline lib/upload.ex]: Better error handling here

File: lib/upload/multi.ex [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+  def download_and_insert_variant(multi, original_blob, variant, transform_fn) do
+    Multi.run(multi, "download_and_insert_#{variant}", fn repo, _ ->
+      with {:ok, blob_path} <- create_random_file(),
+           :ok <- Upload.Storage.download(original_blob.key, blob_path),
+           {:ok, variant_path} <- create_random_file(),
+           :ok <- transform_fn.(blob_path, variant_path, variant),
+           :ok <- cleanup(blob_path),
+           {:ok, blob} <- insert_variant(repo, original_blob, variant, variant_path),
+           :ok <- cleanup(variant_path) do
```
Comment (JohnDoneth) [inline lib/upload/multi.ex]: Can tag each step with a unique error so we can detect which step went wrong. If keeping function syntax use {:error, _} tuple for multi.run return. Add tests for these functions failing. Maybe use a multi.run per step

File: lib/upload/changeset.ex [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+    validate_attachment(changeset, field, :byte_size, fn
+      byte_size when byte_size < max_byte_size -> []
+      _ -> [{field, {message, number: number, unit: unit}}]
+    end)
+  end
+
+  for {unit, multiplier} <- @unit_conversions do
+    defp to_bytes({n, unquote(unit)}), do: n * unquote(multiplier)
+  end
```
Comment (JohnDoneth) [inline lib/upload/changeset.ex:159]: 🤔 Too fancy?

File: lib/upload/blob.ex [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+      nil ->
+        changeset
+
+      original_blob_id ->
+        if repo.get_by(__MODULE__, id: original_blob_id).variant do
+          add_error(
+            changeset,
+            :original_blob_id,
+            "Can not set original_blob_id to a variant blob."
+          )
```
Comment (JohnDoneth) [inline lib/upload/blob.ex]: Use repo.get instead of get_by. Maybe select the variant field only.

File: test/upload_test.exs [hunk 9]
```diff
@@ -1,85 +1,132 @@
...
-             ~r"^logos/[a-z0-9]{32}\.png$"
+  describe "create_multiple_variants/3" do
+    test "create a single variant of an upload" do
+      changeset = change_person(%{avatar: @upload})
+
+      assert {:ok, %{person: person}} = upload_person(changeset)
+      assert person.avatar
+
+      {:ok, [small_variant, small_avif_variant]} =
+        Upload.create_multiple_variants(
```
Comment (JohnDoneth) [inline test/upload_test.exs:66]: One line

File: lib/upload/blob.ex (hunk 1 above)
Comment (JohnDoneth) [reply lib/upload/blob.ex:85]: Added constraint to disallow periods in the key

File: lib/upload.ex (hunk 4 above)
Comment (JohnDoneth) [reply lib/upload.ex]: I'm not exactly sure what was meant by this so I'm leaving it for now.

**Review by ehsieh — APPROVED** [review]

**Pull request merged**